### ISSUE TITLE
DataClassContainsFunctions: operator overloading is flagged

### ISSUE THREAD

**Issue by TWiStErRob** [detekt/detekt#5321]
## Context
Consider having a mathematical/algebraic class, which might have operators. A really simple example:
```kotlin
data class Vector2(val x: Float, val y: Float) {
	operator fun plus(other: Vector2): Vector2 =
		Vector2(x + other.x, y + other.y)
}
```

## Expected Behavior of the rule
Not exactly sure, but it feels like `DataClassContainsFunctions` should allow these. The only other option is writing an extension function:
```kotlin
data class Vector2(val x: Float, val y: Float)
operator fun Vector2.plus(other: Vector2): Vector2 =
	Vector2(x + other.x, y + other.y)
```
which has some trivial drawbacks, for example having to import each function used in math expression at, every, single, usage.

---

Similar, related use case is non-operator functions that do something algebraic (real use case that triggered this issue):
```kotlin
data class SemanticVersion(
	val major: Int,
	val minor: Int,
	val patch: Int,
) : Comparable<AGPVersion> {
	override fun compareTo(other: AGPVersion): Int = // Nothing wrong with this, it's accepted.
		compareValuesBy(this, other, AGPVersion::major, AGPVersion::minor, AGPVersion::patch)

	infix fun compatible(other: SemanticVersion): Boolean = // this is flagged
		this.major == other.major
}
```
again, this could be extracted as extension function.

---

Note: the action coming out of this might be simply updating the documentation, if I misunderstood the intent. (I understand this rule is opt-in, I'm using `allRules`.)

**Comment by BraisGabin** [comment]
To be honest I don't get why this rule should improve your code in general. As long as your data class is immutable it should be completely fine to have methods on it.

I would rather deprecate this rule than "fix this". But an "allow operators" flag could fix this.

As a workaround you could use `conversionFunctionPrefix` and add `plus` but it doesn't feel right...

**Comment by TWiStErRob** [comment]
> But an "allow operators" flag could fix this.

Immediately following on from the main report, notice that it's quite easy to define non-`operator` operator function, so it's quite tricky to define "operators".

**Comment by BraisGabin** [comment]
Yeah, I don't like the "operator" thing either. For that reason, to me, we should deprecate this rule. Theree are a lot of good reasons to define functions inside a `data class`.

I agree that an `Interactor`, `Repository`, `Presenter` or `Mapper` shouldn't be a `data class`. But, if we want to flag that, we should think of more clever ways to detect those cases. Just flag any `data class` with a function feels to broad for a default rule.

**Comment by BraisGabin** [comment]
Could we take a decision on this one?

**Comment by cortinico** [comment]
Having the "allow operators" sounds like a good fix to me.

> Just flag any `data class` with a function feels to broad for a default rule.

That's also why this rule is not enabled by default (and I believe it should stay like that).
@TWiStErRob is using the `allRules` and as such, you get exposed to also those rules that are  considered "controversial" (like this one IMHO).

I'd vote for keeping it, as I understand the value of keeping data classes without behaviour, but also there are valid scenarios where this is not the case.

**Comment by TWiStErRob** [comment]
I also think it's good to keep this, because it would easily catch `data class FooMapper`, which is a true positive finding.

Implement "allow operators" flag works I think for this issue specifically. It might not work for `Comparable` unless the operator-ness inherits in the PSI, but this is an implementation detail.

For algebraic types and functions I think the key is to update the rule documentation, that it might be too broad catch-all, and that there are valid cases for a `@Suppress`. We can easily list some examples.

**Comment by atulgpt** [comment]
Hi, @TWiStErRob / @BraisGabin should I pick this?

**Comment by TWiStErRob** [comment]
Based on the upvotes I think you're good to go. Maybe summarise what your understanding of the requirements as you understand it from above before you start to prevent misunderstandings.

**Comment by atulgpt** [comment]
Hi @TWiStErRob, below are the requirements that I understood

1.  Allow kotlin, `operator fun` via `allowedOperators: List<String>` configuration
2. Maybe? One more config for `allowInfix` configuration.

Below will be proposed rule
```yml
  DataClassContainsFunctions:
    active: true
    conversionFunctionPrefix: [ 'to' ]
    allowedOperators: [ 'plus', 'minus' ]
    allowInfix: false
```

Let me know if this covers all the case

**Comment by TWiStErRob** [comment]
I'm not sure about the list of operators, that complicates things (e.g. have to list all operators by default and maintain that list), but also makes it really flexible. This could be added later should the need arise. I would keep it a simple flag for now.

I personally would accept the infix flag.

@cortinico @BraisGabin thoughts?

**Comment by BraisGabin** [comment]
> have to list all operators by default and maintain that list

I think that, by default, we should keep that list empty so, in terms of maintenance, it should be fine.

Different idea: What do you think about combining `conversionFunctionPrefix` and the operator feature? 

We could deprecate `conversionFunctionPrefix` and create `allowedFunctionNames` or just `allowedFunctions` and have a list of valid names allowing glob patterns. This way you could do:

```yml
allowedFunctions:
  - 'to*'
  - 'plus'
  - 'get'
```

---

If my idea is discarded I would also go with the flag instead of a list of operators. And I would work on the infix on other PR to avoid mix things.

**Comment by TWiStErRob** [comment]
From DX perspective having an empty list by default is horrid. Imagine that every single user would have to configure Detekt in a way that list all operators functions. Each of those developers would have to understand the full list, find it, and encode it in a way that is accepted. Compare this to simplify flipping a flag from false to true. The reason behind this is that operator funs are a class of functions that have very specific semantics and cherry picking might be a bit too granular.

I like the idea of making a more flexible allowlist, but not in scope of this fix.

**Comment by BraisGabin** [comment]
Then go ahead with the flag

**Comment by atulgpt** [comment]
Thanks, @BraisGabin / @TWiStErRob for the confirmation so the final config would be
```yml
  DataClassContainsFunctions:
    active: true
    conversionFunctionPrefix: [ 'to' ]
    allowOperators: false
```
Also, I am considering the default value of `allowOperators` as `false` as currently `allowOperators` will not actually allow only `operator` like functions but rather it will allow all the Kotlin `operator fun` functions

Also suggest the name of the flag, right now I have used `allowOperators` in my example. Still, it could be `allowOperatorFun` for more clarity(as that is what the current implementation will do) and later when we refine this rule we can deprecate old flags. Thoughts?

**Comment by BraisGabin** [comment]
I would go with `allowOperators`. If we need to refine this rule further we should open another issue and talk about it.

**Comment by TWiStErRob** [comment]
Agree with Brais, in the Kotlin naming ["operators" are the function with `operator` keyword](https://kotlinlang.org/docs/operator-overloading.html). Anything else is just "operator-like" function, or "infix functions". There's no clear definition of what constitutes as an operator other than the ones with the keyword.

**Comment by cortinico** [comment]
> I would go with `allowOperators`

+1 for `allowOperators` as a catch-all for any `operator` function

**Comment by atulgpt** [comment]
Hi @BraisGabin / @cortinico  @TWiStErRob, what should be the behaviour of the following code?

```kotlin
interface VerticalLine {
    operator fun get(index: Int): Float
}

class VerticalLineImpl(private val xValue: Float) : VerticalLine {
    override fun get(index: Int): Float {
        return xValue
    }
}

data class VectorOnYAxis(val x: Float) : VerticalLine by VerticalLineImpl(0F)
```

as for normal function, it can be considered as `overridden fun` which is allowed by the rule but in the case of `allowOperators` `false` and `true` what should be desired behaviour?

If I am not mistaken, handling this will require `RequiresTypeResolution`

**Comment by BraisGabin** [comment]
This rule should not care about delegation. And if it should it goes beyond the scope of this issue.

**Comment by TWiStErRob** [comment]
I think that's two different questions:
 * with delegation/inheritance I also think it is a different issue (can `data class` inherit like this?), if you can give a false positive example, maybe open a separate issue to discuss.
 * if you omit `operator` in VerticalLineImpl, then type resolution is required to get the full signature, I think that's unavoidable, right @BraisGabin? See also `Comparable` example above.

**Comment by atulgpt** [comment]
Hi @TWiStErRob for your point 1, currently we are allowing `override fun` inside the data class so that is not an issue. Only problem we will have when someone has set the `allowOperators` as `false` and then using my example still able to add the `operator fun` in the data class.

Although using static analysis we can't guarantee 0 false negatives, so we can also live with the issue as IDK that's very major issue.

BTW code given is compilable and you can play with it at [kotl.in](https://pl.kotl.in/lXK4CJA1c)

**Comment by TWiStErRob** [comment]
Interesting, I think that's just the question of how we define "allowOperators". It looks like "allowOperators: false" is not equivalent to "denyOperators: true", and that's probably ok, just make sure the property description in the config reflects it. Thanks for the clear examples!
